src/vmm/src/lib.rs

```diff
@@ -461,6 +467,143 @@ impl Vmm {
         }
         states
     }
+
+    /// Creates a snapshot of a paused microVM.
+    #[cfg(target_arch = "x86_64")]
+    pub fn create_snapshot(
+        &mut self,
+        params: CreateSnapshotParams,
+    ) -> std::result::Result<(), CreateSnapshotError> {
+        let vcpu_states = self.save_vcpu_states()?;
+
+        let vm_state = self
+            .vm
+            .save_state()
+            .map_err(CreateSnapshotError::SaveVmState)?;
+
+        let device_states = self.save_mmio_device_states();
+
+        let mem_size_mib =
+            self.guest_memory()
+                .map_and_fold(0, |(_, region)| region.len(), |a, b| a + b)
+                >> 20;
+
+        let microvm_state = MicrovmState {
+            vm_info: VmInfo { mem_size_mib },
+            vm_state,
+            vcpu_states,
+            device_states,
+        };
+
+        self.snapshot_memory_to_file(&params.mem_file_path, mem_size_mib, params.snapshot_type)?;
+
+        let mut snapshot_file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(params.snapshot_path)
+            .map_err(CreateSnapshotError::SnapshotBackingFile)?;
+
+        // TODO: fallback to current version.
+        let version_map = VersionMap::new();
+        let version = match params.version {
+            Some(version) => version,
+            _ => 1,
```

Comment:
Why not do this instead of hardcoding it ?
```suggestion
            _ => version_map.latest_version(),
```